### PR TITLE
optimize (b ? c1 : c2) to (c1 + b *(c2 - c1))

### PR DESCRIPTION
**Pull request by WalterBright**
When `b` will evaluate to 0 or 1 and (c2-c1) is a power of 2.

### CODE REVIEW

File: src/ddmd/backend/cgelem.c [hunk 1]
```diff
@@ -1900,20 +1900,65 @@ STATIC elem * elcond(elem *e, goal_t goal)
...
+                                    continue;
+                                default:
+                                    assert(0);
+                            }
+                            break;
+                        }
+                        if (i1 < i2)
+                        {
+                            ec2->EV.Vllong = i2 - i1;
+                            e1 = el_bin(OPxor,e1->Ety,e1,el_long(e1->Ety,1));
```
Comment (yebblies) [inline src/ddmd/backend/cgelem.c]: This isn't covered by the comment above...

File: src/ddmd/backend/cgelem.c [hunk 2]
```diff
@@ -1900,20 +1900,65 @@ STATIC elem * elcond(elem *e, goal_t goal)
...
-                /*   replace ((a & b) ? 0 : b) with ((a & b) ^ b)       */
-                if (e1->Eoper == OPand && e1->E2->Eoper == OPconst &&
+                if ((ty1 == TYbool && !OTlogical(e1->Eoper) || e1->Eoper == OPand && e1->E2->Eoper == OPconst) &&
                     tysize(ty) == tysize(ec1->Ety))
                 {
-                    b = el_tolong(e1->E2);
-                    if (ispow2(b) != -1)        /* if only 1 bit is set */
+                    targ_llong b = ty1 == TYbool ? 1 : el_tolong(e1->E2);
+
+                    if (b == 1 && ispow2(i1 - i2))
```
Comment (yebblies) [inline src/ddmd/backend/cgelem.c]: This requirement needs to be in the comment too

File: src/ddmd/backend/cgelem.c [hunk 3]
```diff
@@ -1900,20 +1900,65 @@ STATIC elem * elcond(elem *e, goal_t goal)
...
+                                    sz = 4;
+                                    continue;
+                                case 4:
+                                    e1 = el_una(OPu32_64, TYullong, e1);
+                                    sz = 8;
+                                    continue;
+                                default:
+                                    assert(0);
+                            }
+                            break;
```
Comment (yebblies) [inline src/ddmd/backend/cgelem.c]: Unreachable?

File: src/ddmd/backend/cgelem.c [hunk 4]
```diff
@@ -1953,14 +1998,14 @@ STATIC elem * elcond(elem *e, goal_t goal)
                 // code for without using jumps.
 
                 // Try to replace (!e1) with (e1 < 1)
-                else if (e1->Eoper == OPnot && !OTrel(e1->E1->Eoper))
+                else if (e1->Eoper == OPnot && !OTrel(e1->E1->Eoper) && e1->E1->Eoper != OPand)
```
Comment (yebblies) [inline src/ddmd/backend/cgelem.c]: What is this for?

**Review by yebblies — COMMENTED** [review]
Also, why does this only apply when a - b is a power of two?  Isn't the transform valid if they're not?

File: src/ddmd/backend/cgelem.c (hunk 2 above)
Comment (yebblies) [inline src/ddmd/backend/cgelem.c]: Also, why is this checking `ispow2` is non-zero, while below it checks `ispow2(b) != -1`?

**Comment by WalterBright** [comment]
> why does this only apply when a - b is a power of two?

Because then it will be transformed later into a shift.

> Isn't the transform valid if they're not?

Yes, but the multiply is more expensive than the ?:, but the shift is cheaper.

File: src/ddmd/backend/cgelem.c (hunk 4 above)
Comment (WalterBright) [inline src/ddmd/backend/cgelem.c]: If it's an `and` operation, the optimization is a pessimization.

**Comment by yebblies** [comment]
Then why not emit a shift directly?  The description of this optimization is misleading since you're not actually transforming to a multiply.

File: src/ddmd/backend/cgelem.c (hunk 1 above)
Comment (WalterBright) [inline src/ddmd/backend/cgelem.c]: fixed

File: src/ddmd/backend/cgelem.c (hunk 3 above)
Comment (WalterBright) [inline src/ddmd/backend/cgelem.c]: made better

File: src/ddmd/backend/cgelem.c (hunk 2 above)
Comment (WalterBright) [inline src/ddmd/backend/cgelem.c]: good catch

**Comment by WalterBright** [comment]
> Then why not emit a shift directly?

Because this is slightly simpler and easier to understand. Quite a bit of the transformations done rely on further transformations.

> The description of this optimization is misleading since you're not actually transforming to a multiply.

Yes, you are, at this point in the optimization.

**Pull request merged**